daemon/algod/api/server/v2/dryrun.go

```
@@ -0,0 +1,295 @@
+	// see go-algorand/data/transactions/logic/debugger.go refreshDebugState() for which fields are updated
+	ds := logic.DebugState{
+		PC:      state.PC,
+		Line:    state.Line,
+		Error:   state.Error,
+		Stack:   state.Stack,
+		Scratch: state.Scratch,
+	}
+	ddr.history = append(ddr.history, ds)
+	ddr.updateScratch()
+	return nil
+}
+
+// Complete is called when the program exits (DebuggerHook interface)
+func (ddr *dryrunDebugReceiver) Complete(state *logic.DebugState) error {
+	return ddr.Update(state)
+}
+
+// LedgerForLogic
+type dryrunLedger struct {
+	dr *DryrunRequest
+}
+
+// LedgerForLogic
+func (dl *dryrunLedger) Balance(addr basics.Address) (basics.MicroAlgos, error) {
+	return basics.MicroAlgos{Raw: 0}, nil
+}
+func (dl *dryrunLedger) Round() basics.Round {
+	return basics.Round(dl.dr.Round)
+}
+func (dl *dryrunLedger) AppGlobalState(appIdx basics.AppIndex) (basics.TealKeyValue, error) {
+	for _, app := range dl.dr.Apps {
+		if app.AppIndex == uint64(appIdx) {
+			return app.Params.GlobalState, nil
+		}
+	}
+	return nil, nil
+}
+func (dl *dryrunLedger) AppLocalState(addr basics.Address, appIdx basics.AppIndex) (basics.TealKeyValue, error) {
+	for _, st := range dl.dr.AccountAppStates {
+		if appIdx == st.AppIndex && addr == st.Account {
+			return st.State.KeyValue, nil
+		}
+	}
+	return nil, nil
+}
+func (dl *dryrunLedger) AssetHolding(addr basics.Address, assetIdx basics.AssetIndex) (basics.AssetHolding, error) {
+	for _, acct := range dl.dr.Accounts {
+		// TODO: check all address decodes when we first receive the dryrun request
+		xaddr, err := basics.UnmarshalChecksumAddress(acct.Address)
+		if err != nil {
+			continue
+		}
+		if xaddr == addr {
+			for _, ah := range *acct.Assets {
+				if ah.AssetId == uint64(assetIdx) {
+					return basics.AssetHolding{Amount: ah.Amount, Frozen: ah.IsFrozen}, nil
+				}
+			}
+		}
+	}
+	return basics.AssetHolding{Amount: 0, Frozen: false}, nil
```

This should return an error if the holding isn't found